### PR TITLE
[5356] Reduce the failure rate for low-level IOCP port listen test.

### PR DESCRIPTION
**Pull request by adiroiban**
## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/5356
* [x] I ran `tox -e black-reformat` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [x] I ran additional checks listed at [Getting Your Patch Accepted](https://twistedmatrix.com/trac/wiki/TwistedDevelopment#GettingYourPatchAccepted)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [x] I have updated the automated tests.
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.

### CODE REVIEW

File: src/twisted/internet/test/test_iocp.py
Comment (mthuurne) [inline src/twisted/internet/test/test_iocp.py]: Typos: "excepted" -> "expected" and "retying" -> "retrying".

File: src/twisted/internet/test/test_iocp.py
Comment (mthuurne) [inline src/twisted/internet/test/test_iocp.py]: The `pass` is redundant.

File: src/twisted/internet/test/test_iocp.py
Comment (mthuurne) [inline src/twisted/internet/test/test_iocp.py]: I don't understand; what is that "second try"?

**Comment by adiroiban** [comment]
This is ready to review...

So... to minimize the test run time... it will try to call setsockopt right away and then wait only on errors.

File: src/twisted/internet/iocpreactor/tcp.py
Comment (mthuurne) [inline src/twisted/internet/iocpreactor/tcp.py]: While that URL still works, it now redirects to:
https://docs.microsoft.com/en-us/windows/win32/api/mswsock/nf-mswsock-acceptex

File: src/twisted/internet/test/test_iocp.py
Comment (graingert) [inline src/twisted/internet/test/test_iocp.py]:
```suggestion
            # We make three attempts because two are not enough.
```

Is this from experiment? or just a guess?

File: src/twisted/internet/test/test_iocp.py
Comment (graingert) [inline src/twisted/internet/test/test_iocp.py]: socket.error is a deprecated alias for `OSError`

**Comment by adiroiban** [comment]
Many thanks @graingert and @mthuurne for your review.

I am not convince that this will fix the error :( ...as I saw another error on ipv4. Might need to retry more than 2 times.

But I will try to trigger more tests and see how it goes

**Comment by adiroiban** [comment]
@graingert and @mthuurne please give it another try.

I hope that this will reduce the failure rate for this test.
I expect the test to still fail from time to time, and when it happened the number of retries can be increased.

Thanks!

File: src/twisted/internet/test/test_iocp.py
Comment (graingert) [inline src/twisted/internet/test/test_iocp.py:111]: won't this socketError get deleted?

```python
def demo():
    e = None
    try:
        raise Exception
    except Exception as e:
        pass

    return e


demo()
```

```pytb
Traceback (most recent call last):
  File "demo.py", line 11, in <module>
    demo()
  File demo.py", line 8, in demo
    return e
UnboundLocalError: local variable 'e' referenced before assignment
```

Comment (adiroiban) [reply src/twisted/internet/test/test_iocp.py:111]: OMG. Python 3 is crazy :) True

Sorry, I am still working only on Python2.

But why was this error not picked up by the linters.

it looks like a common error to be make when migrating from py2 to py3.

thanks @graingert . Great feedback :)

Comment (graingert) [reply src/twisted/internet/test/test_iocp.py:111]:
```python
for attempts_remaining in range(4, -1, -1):
    try:
        server.setsockopt(SOL_SOCKET, SO_UPDATE_ACCEPT_CONTEXT, pack("P", port.fileno()))
    except OSError as e:
        if attempts_remaining == 0 or e.errno != getattr(errno, "WSAENOTCONN"):
            raise
        time.sleep(0.2)
```

Comment (adiroiban) [reply src/twisted/internet/test/test_iocp.py:111]: will fix this.

More info https://stackoverflow.com/a/44258937/539264 and https://www.python.org/dev/peps/pep-3110/

File: src/twisted/internet/test/test_iocp.py
Comment (graingert) [inline src/twisted/internet/test/test_iocp.py]: I think it's neater to count backwards, than assign and re-assign the error:

```python
for attempts_remaining in reversed(range(5)):
    try:
        server.setsockopt(SOL_SOCKET, SO_UPDATE_ACCEPT_CONTEXT, pack("P", port.fileno()))
        break
    except OSError as e:
        if attempts_remaining == 0 or e.errno != getattr(errno, "WSAENOTCONN"):
            raise
        time.sleep(0.2)
```

Comment (adiroiban) [reply src/twisted/internet/test/test_iocp.py]: yep. Will do it. thanks :)

**Comment by adiroiban** [comment]
I think that this is ready for a new round of reviews.

5 years ago when I tested the initial code I think that I have temporary update the SupportTests to call the ipv4 and ipv6 accept tests 100 times in a row in a single test run.

I think that I can also do it now to see that if the tests is executed 100 or 1000 times in the Azure or GitHub VMs they will not fail.

File: src/twisted/internet/test/test_iocp.py
Comment (mthuurne) [inline src/twisted/internet/test/test_iocp.py]:
```suggestion
        These are a flaky tests.
```

File: src/twisted/internet/test/test_iocp.py
Comment (mthuurne) [inline src/twisted/internet/test/test_iocp.py]:
```suggestion
            # and IPV6 with "[Errno 10057] A request to send or receive ..."
```

File: src/twisted/internet/test/test_iocp.py
Comment (mthuurne) [inline src/twisted/internet/test/test_iocp.py:115]: I think it would be clearer to check `errno` before `attemptsRemaining`. The `errno` check is like a subtype check: we don't want to catch all `OSError` instances, but only one particular kind. So it would make sense to have it immediately following the `except OSError` line.

File: src/twisted/internet/test/test_iocp.py
Comment (mthuurne) [inline src/twisted/internet/test/test_iocp.py]: `reversed(range(5))` produces the sequence (4, 3, 2, 1, 0), so there is no way `attemptsRemaining` is ever below zero.

File: src/twisted/internet/test/test_iocp.py
Comment (mthuurne) [inline src/twisted/internet/test/test_iocp.py:95]: Since the attempts + delay regulate the number of unintended failures, why not set the values a bit higher? For example a 10 second or even 30 second timeout is fine, since that worst case is only hit when the test fails and even then only when it fails in a very particular way (`accept()` succeeding but the socket never reaching connected state).

**Review by mthuurne — CHANGES_REQUESTED** [review]
I added some more comments. Most are minor, but the `attemptsRemaining < 0` one looks like a bug.

File: src/twisted/internet/test/test_iocp.py
Comment (adiroiban) [reply src/twisted/internet/test/test_iocp.py:95]: I think that this test should not take more than 5 seconds to executes.

If it takes 10 seconds for an IOCP socket to accept an already triggered connection, then I think that something is very wrong.

Most of the time this is an instant operation.... so anything about 5 second should raise a warning.

I wanted to keep lower values to see how often it fails with a timeout below 1 second.

File: src/twisted/internet/test/test_iocp.py
Comment (adiroiban) [reply src/twisted/internet/test/test_iocp.py]: True. Stupid me :) Thanks

File: src/twisted/internet/test/test_iocp.py
Comment (mthuurne) [reply src/twisted/internet/test/test_iocp.py:95]: A test case taking unusually long before passing can indeed be an indication of a deeper problem, but if we are interested in measuring that, it would be better to make that part of the test framework rather than individual tests, in my opinion.

I have no data to support my gut feeling that a 10 seconds timeout is better than 1 second, so let's leave it as it is and see what happens.

**Review by mthuurne — APPROVED** [review]
Looks good now. Let's hope our theories are correct and this indeed fixes the flaky tests.

**Comment by adiroiban** [comment]
Thanks @mthuurne  for the review.

I pushed another commit just to test that all is ok.

There are many comments, but I tried to summarize all the information for the PR conversation.

Let me know if it make sense.

Thanks!

File: src/twisted/internet/test/test_iocp.py
Comment (mthuurne) [inline src/twisted/internet/test/test_iocp.py]:
```suggestion
        A port starts listening (is bound) at the low-level socket without
```

File: src/twisted/internet/test/test_iocp.py
Comment (mthuurne) [inline src/twisted/internet/test/test_iocp.py]:
```suggestion
        calling accept() yet.
```

File: src/twisted/internet/test/test_iocp.py
Comment (mthuurne) [inline src/twisted/internet/test/test_iocp.py]:
```suggestion
        After the client is connected IOCP accept() is called, which is the
```

File: src/twisted/internet/test/test_iocp.py
Comment (mthuurne) [inline src/twisted/internet/test/test_iocp.py]:
```suggestion
        target of these tests.
```

File: src/twisted/internet/test/test_iocp.py
Comment (mthuurne) [inline src/twisted/internet/test/test_iocp.py]:
```suggestion
        the socket is not ready right away after calling IOCP accept().
```

File: src/twisted/internet/test/test_iocp.py
Comment (mthuurne) [inline src/twisted/internet/test/test_iocp.py]:
```suggestion
        It should not take more than 5 seconds for a socket to be ready, as
```

**Comment by adiroiban** [comment]
We have another failure on IOCP...but is not related to the scope of this PR.

https://dev.azure.com/twistedmatrix/twisted/_build/results?buildId=2813&view=logs&j=103d3058-bd81-5adc-9d09-335c8b1168a8&t=c2351c64-3c18-5e69-f4d7-151e9583a5ee&l=15866

```
Traceback (most recent call last):
Failure: twisted.trial.util.DirtyReactorAggregateError: Reactor was unclean.
Selectables:
<ThrottlingProtocol #1 on 52211>

twisted.test.test_policies.ThrottlingTests.test_limit
```

**Comment by mthuurne** [comment]
Is that change in `policies.ThrottlingFactory` related to the IOCP accept test? Because at a first glance it seems unrelated, in which case it would be better to address it in a separate PR.

**Comment by adiroiban** [comment]
Hi.  Thanks for the review. You are right.

I will revert it.

 I just wanted to test the code and trigger another set of tests for IOCP. 

With tests on Azure I can't retry a failed test so I had to commit something.

I will revert the change and if it will pass I will merge it and open another ticket for the other test.

File: src/twisted/internet/test/test_iocp.py
Comment (graingert) [inline src/twisted/internet/test/test_iocp.py]:
```suggestion
                    # We ran out of retries.
```

```suggestion
                    # We ran out of attempts.
```

**Review by graingert — APPROVED** [review]

**Comment by mthuurne** [comment]
> With tests on Azure I can't retry a failed test so I had to commit something.

That's strange, because I do get the option to re-run failed tests with Azure. You have to click through to Azure's web front-end though; you can't trigger it from GitHub. And while the re-run link is shown as soon as a test fails, it only seems to actually start re-running things if clicked when all tasks within a run have completed.

_Edit:_ Actually, I misremembered: the re-run link is in the sidebar on GitHub and **not** on the Azure site.

**Comment by adiroiban** [comment]
@mthuurne thanks for the Azure re-run link :) I was searching for it on the Azure Devops page:)

But I hope that this IOCP test is now fixed for good :)

I will wait for all the tests to be green and will merge ...then we can see if iocp test continues to fail

**Comment by graingert** [comment]
@adiroiban it passed!

**Comment by adiroiban** [comment]
@mthuurne please see latest commit and let me know if that drive-by comment is ok.

For the non-related flaky test discovered during this PR, I have created a separate ticket but add the link to the ticket as a drive by in this PR.

Do you still approve this PR?

Thanks!

**Comment by mthuurne** [comment]
Yes, I approve of its current state.

**Comment by wsanchez** [comment]
Coolness.  Thanks for working on this, Adi.

**Comment by adiroiban** [comment]
After the merge... many tests failed. @mthuurne  I see the "re-run" link/button, but it looks like it is not doing anything.

**Comment by mthuurne** [comment]
> After the merge... many tests failed. @mthuurne I see the "re-run" link/button, but it looks like it is not doing anything.

When I click the "re-run" link, I see "You have successfully requested twisted.twisted be rerun." appear at the top of the screen, but nothing starts re-running or even leaves the error state. I don't have any ideas why or how to debug this, unfortunately.

**Comment by adiroiban** [comment]
@wsanchez this is where you can see all the checks for trunk 

image

**Pull request merged**
